### DISCUSSION TITLE
Built-in Raven Event Store

### DISCUSSION THREAD

**Discussion by kblooie** [ravendb/ravendb#18041] [Ideas]
I think that Raven should offer first-class support for a distributed Event Store, like the built-in solution for the MartenDB (https://martendb.io/events/).

Raven has all the building blocks for a great event store implementation, but it's not trivial to build. Also, it's not trivial to provide solutions to manage all the operational features needed to support an event store over time. It would be awesome to have first class support for delivering and managing projections like indexes. This would also align nicely w/ the broker ETL features already in raven.

**Comment by ayende** [comment]
Give your familiarity with RavenDB, how do you think the design for this feature look like?

**Comment by kblooie** [comment]
Sure. Fair warning, I've just started thinking about this today.

## Event Store

Marten's implementation is pretty straight forward:

https://github.com/JasperFx/marten/tree/master/src/Marten/Events

I think splitting into 2 models makes sense:

### Stream

Store metadata about version & other stream metadata. 

* Should be a light-weight payload
* Stored either as Compare Exchange or as a document with Atomic Transactions on the StreamId

### Events

Each event should be stored as a separate documents for easy access & indexing. I think they should be stored in a document database to make them easy to index and store along w/ projected view documents.

ID: <stream_id>/<stream_version>

Event contract is pretty standard: 

https://martendb.io/events/storage.html#event-metadata-in-code

#### Ordering Across Streams

The primary complexity that I've struggled with in previous Event Store attempts is a way to be able to replay events across all streams **in order**, while keeping storage distributed & "fast" in Raven.

In Marten, it's implemented as Sequence (long) "A global sequential number identifying the Event".

It doesn't have to be a sequential value, there just needs to be a way to replay events in order across streams.

This has never been a major issue for me, and I usually don't need consistency with these types of replays. However, it would be nice not to have to think about different consistency guarantees on ordering in different types of projections.

### First Class Support - Guards To Prevent Stupid

It would be really nice if Raven would have built-in guards that prevent Events and Stream models from being modified after storage:

* Make it hard to modify event bodies to ensure that it's intentional
* Make it impossible to delete an event from the stream
* Don't allow storing event correlations that don't make sense for the stream

These are things a library can't do.

## Projections

Regardless of Single Stream vs Multi-Stream projections, the main thing to look at is when they're executed.

### Inline Projections

Saved in the same transaction as writing to the stream:

* Only supports single stream projections
* Conflicts
  * Projected documents shouldn't be modified outside of the transaction, so conflicts isn't a huge risk
  * Conflicts can be resolved by replaying the stream

### Async Projections

Processed outside stream transactions. This fits really nicely into subscriptions.

### First Class - Index Projections

This is the differentiating feature for me. Being able to implement projections as indexes & have them run in raven is a game changer.

* Dead simple to author
* Can easily support different event versions without needing to maintain c# types
* Using Management Studio for op support and metrics
* No infrastructure to mess with
* No Conflicts? Yes please!

Index projections need to support saving denormalized documents that can be loaded via a stream id for single stream projections.

### First Class - Live Projections

Being able to ad-hoc create projections in the Management Studio is also a game changer for operational support.

## Other First Class Features

### Integration With Broker ETL

Being able to publish stored events to a RabbitMQ without needing to manage a separate process is huge.

### Management Studio Operational Dashboard

Be able to have insight into your Event Store:

* Easily see all events for a stream
* Stream Archival Tasks
* Projection performance metrics

**Comment by ayende** [comment]
Stream in this case is the thing of which we have events?

In the context of an online store, a stream would be a shopping cart with the events being item-added, item-removed ?

* Why do you think it should be globally consistent? If they are common, that adds expensive operation, and I don't think that the consistency demands it.

* Ordering only matters _within_ a stream, no? What is the scenario that you have for ordering between streams? The problem with sequence number is a distributed system is that you are going to run into _nasty_ issues at the edge.

Two separate instances cannot easily agree on order of events without coordination, and for performance, we want to avoid that.

*Preventing deletions* - in general - is actually a problem.

We have something similar (you cannot easily delete revisions), and it turns out that there are good reasons why you want that.
* Removal of data because of GDPR etc
* Changing the event schema after an upgrade, etc.
* Accidently put sensitive data like plain text password

* What do you mean, "event correlation that don't make sense" - not sure what this means.

*Inline projections* - what happened when you have concurrent writes to the stream on two nodes at the same time. Each has only partial data, after all. And the projection document would conflict.

*Index projections* - Did you see: https://ayende.com/blog/posts/series/185729-A/data-modeling-with-indexes

*Live projections* - that may use the test index feature.

**Comment by karmeli87** [comment]
Hi,

Actually, currently we are working on something very similar. A persistence plug-in for Akka.Net.

To keep it performant we avoid the cluster-wide operations as much as possible, no cluster-wide, compare exchange or even subscriptions.

Each stream has a unique ID and each Event has the following ID format `<stream_id>/<lexical_seq_num>` e.g. `Stream/Foo/0000000000000000001` the leading zeros ensure us that all events for a given stream are lexically sorted.

To ensure valid state across the nodes we check that the previous event has the `sequence_number - 1`, this way we can't have gaps in the events for a given stream.

Yes we can have lagging nodes, but replaying from each node will bring the system to a valid state.

Akka.Net has the notion of tagging, each event can be tagged and query upon this tag later, I guess Projections here are more generic and powerful.

Indexing those tags is done with the RavenDB indexes, but querying them is done with "continues query", we periodically check if there are new events. 

The challenge here is to have a distributed offset, to continue from the point where the previous query ended, we solve that by indexing the change vector for each event.

You can check the code here `https://github.com/karmeli87/Akka.Persistence.RavenDB` please keep in mind that it is still very preliminary implementation as we just started to work on it.

Personally I really like the idea of having Events as part RavenDB itself, maybe as an additional package.

Any feedback or ideas are more than welcome :-)

**Comment by kblooie** [comment]
_Stream in this case is the thing of which we have events?_

Yes, it's a consistency boundary around events. It usually has a version, which consists of the number of events in the stream.

Most of the time you want to be consistent when appending to a stream. Many uses of event sourcing do an in-memory replay of past events to make domain decisions & throw concurrency errors that are handled when another request has modified the stream.

In the context of an online store, a stream would be a shopping cart with the events being item-added, item-removed?

Yes, ShoppingCart is a common example, though it's lifetime is usually really short.

_Why do you think it should be globally consistent? If they are common, that adds expensive operation, and I don't think that the consistency demands it._

I disagree. It is not trivial to reconcile 2 stream histories. In many domains, I'd rather take an availability / performance hit than try to reconcile conflicts across event histories.

My 2nd option where the Stream is a document along with using atomic guards on the StreamId would allow you to opt into cluster transactions.

_Ordering only matters within a stream, no? What is the scenario that you have for ordering between streams? The problem with sequence number is a distributed system is that you are going to run into nasty issues at the edge._

Composite Views

In most cases Aggregated views are where ordering doesn't matter, but there are times where you want to build a composite view from multiple streams. You want replay to be consistent if you ever reset the projection.

You can take a look at a few examples here:

https://martendb.io/events/projections/multi-stream-projections.html

Order Submission Events with Shipping Events for the order correlated on OrderId.  Pardon my crappy example, but: 

You mark an item as approved for shipping and that's processed in the projection before the item was added in an order stream. That's complex and it's sucks to deal with.

_Two separate instances cannot easily agree on order of events without coordination, and for performance, we want to avoid that._

I absolutely agree that a sequential number is not the way to go. We just need to be able to be consistent in sequence. Could a time sortable id be enough here? Something like twitter / instagram's solutions using nodeids w/ time to generate a time sortable id?

_Preventing deletions - in general - is actually a problem. We have something similar (you cannot easily delete revisions), and it turns out that there are good reasons why you want that._

I get it. I'm really just wanting guards to prevent stupid. Something like an override that needs to be set if a change / deletion is detected that's not part of an archive or warnings in the management studio. 

Revisions do add a safety net as long as they're turned on and the history / deletion rules apply. I'd like some safe defaults here at least.

_Changing the event schema after an upgrade, etc._

I think Marten's article on Event Versioning is a good one:

https://martendb.io/events/versioning.html

You're really not supposed to change history. Rather, prefer to be additive. A "fix" could lose the reason why the "Next" event happened due to a decision based on the data "pre-fix". However, it happens. That's why I said "Hard" and not "Impossible".

I don't want some random script accidentally screwing w/ event history. I'd like protections.

_What do you mean, "event correlation that don't make sense" - not sure what this means._

Someone makes a change to the stream to set the version to 9 when there are 10 events. Someone deletes event 7, but not 1-6. Now the stream can't be replayed for write decisions using standard logic.

_Inline projections - what happened when you have concurrent writes to the stream on two nodes at the same time. Each has only partial data, after all. And the projection document would conflict._

Not if you're using a cluster transaction as the Raft vector always wins. If you use inline then only the projection should modify the document, which is why you can only support single-stream projections.

_Index projections - Did you see: https://ayende.com/blog/posts/series/185729-A/data-modeling-with-indexes_

I did. I'll re-read this weekend, but I remember thinking that the use case would only fit a small number of projections. I also would want 1st class support for writing projection indexes that has a client API similar to Marten's projection API:

https://martendb.io/events/projections/
https://martendb.io/events/projections/multi-stream-projections.html#simple-example-of-events-updating-multiple-views

_Live projections - that may use the test index feature._

Yup, but having 1st class support w/ a UI that easily brings things together is what would make things stand out.

**Comment by ayende** [comment]
Sorry for the delay in answering.

Regarding consistency - there is an issue here about ordering and considerations. In the shopping cart example, the order of events don't matter (until you have the Pay event, of course).

For longer term sequences, I can think of examples where the ordering matters or not. That leads to another important aspect. What is the _size_ of each stream? Are we expecting a stream to have tens of thousands of events? Or top at a few hundreds?

What level of concurrency is actually expected?

Having an option to make that cluster-wide tx or not is really useful, and _not_ requiring cluster-wide is likely to be an important design consideration.


Time sortable ids are trivial to do, yes. The problem is that you run into _what_ time? If it is sufficent to be roughly sorted, that is easy. But consider the scenario where you have two requests going to two nodes at roughly the same time. Do you care about their ordering?

Preventing deletions - that can be something like `@prevent-delete: true` marker in the metadata, which will simply reject deletes (but you can remove that and then delete). The problem is that those code-paths have been _heavily_ optimized, and adding a check is something that I don't want to do. 
More importantly, we run into really complex question with: "deletion came from replication, what now?", etc.

> Someone deletes event 7, but not 1-6.

That would require building a _lot_ of insight into RavenDB about how those are correlated to one another. 
If we assume that you are talking about adding those protections at the database layer, not just the API one.
